# 0. Initialization

In [109]:
import math
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [110]:
df = pd.read_csv('../data/Improving Employee Retention by Predicting Employee Attrition Using Machine.csv', sep=',')
df.head()

Username  EnterpriseID StatusPernikahan JenisKelamin  \
0  spiritedPorpoise3        111065    Belum_menikah         Pria   
1    jealousGelding2        106080    Belum_menikah         Pria   
2      pluckyMuesli3        106452          Menikah         Pria   
3   stressedTruffle1        106325    Belum_menikah         Pria   
4        shyTermite7        111171    Belum_menikah       Wanita   

  StatusKepegawaian                      Pekerjaan           JenjangKarir  \
0         Outsource   Software Engineer (Back End)  Freshgraduate_program   
1          FullTime                   Data Analyst  Freshgraduate_program   
2          FullTime  Software Engineer (Front End)  Freshgraduate_program   
3         Outsource  Software Engineer (Front End)  Freshgraduate_program   
4          FullTime                Product Manager  Freshgraduate_program   

  PerformancePegawai     AsalDaerah     HiringPlatform  ...         NomorHP  \
0       Sangat_bagus  Jakarta Timur  Employee_Referral  ...  +6282232522xxx   
1      Sangat_kurang  Jakarta Utara            Website  ...  +6281270745xxx   
2              Bagus  Jakarta Timur             Indeed  ...  +6281346215xxx   
3              Bagus  Jakarta Pusat           LinkedIn  ...  +6283233846xxx   
4              Bagus  Jakarta Timur           LinkedIn  ...  +6287883263xxx   

                             Email  TingkatPendidikan  PernahBekerja  \
0   spiritedPorpoise3135@yahoo.com           Magister              1   
1     jealousGelding2239@yahoo.com            Sarjana              1   
2      pluckyMuesli3961@icloud.com           Magister              1   
3  stressedTruffle1406@hotmail.com            Sarjana              1   
4         shyTermite7149@gmail.com            Sarjana              1   

   IkutProgramLOP   AlasanResign TanggalLahir TanggalHiring  \
0             1.0  masih_bekerja   1972-07-01    2011-01-10   
1             1.0  toxic_culture   1984-04-26    2014-01-06   
2             1.0      jam_kerja   1974-01-07    2011-01-10   
3             0.0  masih_bekerja   1979-11-24     2014-2-17   
4             0.0    ganti_karir   1974-11-07    2013-11-11   

  TanggalPenilaianKaryawan  TanggalResign  
0                2016-2-15              -  
1                2020-1-17      2018-6-16  
2               2016-01-10      2014-9-24  
3               2020-02-04              -  
4                2020-1-22     2018-09-06  

[5 rows x 25 columns]

# 1. Data Preprocessing

## 1.1 Quick EDA

In [111]:
# Informasi dataset
print("Info Dataset:")
print(f"Jumlah baris : {df.shape[0]}")
print(f"Jumlah kolom : {df.shape[1]}")

Info Dataset:
Jumlah baris : 287
Jumlah kolom : 25


In [112]:
# Informasi kolom
print("Info Kolom:")
print(df.info())

Info Kolom:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 287 entries, 0 to 286
Data columns (total 25 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Username                            287 non-null    object 
 1   EnterpriseID                        287 non-null    int64  
 2   StatusPernikahan                    287 non-null    object 
 3   JenisKelamin                        287 non-null    object 
 4   StatusKepegawaian                   287 non-null    object 
 5   Pekerjaan                           287 non-null    object 
 6   JenjangKarir                        287 non-null    object 
 7   PerformancePegawai                  287 non-null    object 
 8   AsalDaerah                          287 non-null    object 
 9   HiringPlatform                      287 non-null    object 
 10  SkorSurveyEngagement                287 non-null    int64  
 11  SkorKepuasanPegawai              

In [113]:
numerical_features = ['EnterpriseID', 'SkorSurveyEngagement', 'SkorKepuasanPegawai', 
                      'JumlahKeikutsertaanProjek', 'JumlahKeterlambatanSebulanTerakhir',
                      'JumlahKetidakhadiran', 'IkutProgramLOP'
                      ]

# categorical_features = df.select_dtypes(include=[object])
categorical_features = ['StatusPernikahan', 'JenisKelamin', 'StatusKepegawaian', 'Pekerjaan', 
                        'JenjangKarir', 'PerformancePegawai', 'AsalDaerah', 'HiringPlatform',
                        'TingkatPendidikan', 'PernahBekerja', 'AlasanResign'
                        ]

In [114]:
# Statistical summaries feature numerical
df[numerical_features].describe()

EnterpriseID  SkorSurveyEngagement  SkorKepuasanPegawai  \
count     287.000000            287.000000           282.000000   
mean   105923.324042              3.101045             3.904255   
std      4044.977599              0.836388             0.913355   
min    100282.000000              1.000000             1.000000   
25%    101269.000000              3.000000             3.000000   
50%    106069.000000              3.000000             4.000000   
75%    110514.500000              4.000000             5.000000   
max    111703.000000              5.000000             5.000000   

       JumlahKeikutsertaanProjek  JumlahKeterlambatanSebulanTerakhir  \
count                 284.000000                          286.000000   
mean                    1.179577                            0.412587   
std                     2.294441                            1.275016   
min                     0.000000                            0.000000   
25%                     0.000000                            0.000000   
50%                     0.000000                            0.000000   
75%                     0.000000                            0.000000   
max                     8.000000                            6.000000   

       JumlahKetidakhadiran  IkutProgramLOP  
count            281.000000       29.000000  
mean              10.448399        0.517241  
std                6.902252        0.508548  
min                1.000000        0.000000  
25%                5.000000        0.000000  
50%               10.000000        1.000000  
75%               15.000000        1.000000  
max               55.000000        1.000000

In [115]:
# Statistical summaries feature categorical
df[categorical_features].describe()

StatusPernikahan JenisKelamin StatusKepegawaian  \
count               287          287               287   
unique                5            2                 3   
top       Belum_menikah       Wanita          FullTime   
freq                132          167               217   

                           Pekerjaan           JenjangKarir  \
count                            287                    287   
unique                            14                      3   
top     Software Engineer (Back End)  Freshgraduate_program   
freq                             109                    169   

       PerformancePegawai     AsalDaerah HiringPlatform TingkatPendidikan  \
count                 287            287            287               287   
unique                  5              5              9                 3   
top                 Biasa  Jakarta Pusat         Indeed           Sarjana   
freq                   85             72             85               176   

       PernahBekerja   AlasanResign  
count            287            221  
unique             2             11  
top                1  masih_bekerja  
freq             286            132

In [116]:
# Pengecekan data yang hilang
print("\nPengecekan Missing Values:")
df.isnull().sum()


Pengecekan Missing Values:


Username                                0
EnterpriseID                            0
StatusPernikahan                        0
JenisKelamin                            0
StatusKepegawaian                       0
Pekerjaan                               0
JenjangKarir                            0
PerformancePegawai                      0
AsalDaerah                              0
HiringPlatform                          0
SkorSurveyEngagement                    0
SkorKepuasanPegawai                     5
JumlahKeikutsertaanProjek               3
JumlahKeterlambatanSebulanTerakhir      1
JumlahKetidakhadiran                    6
NomorHP                                 0
Email                                   0
TingkatPendidikan                       0
PernahBekerja                           0
IkutProgramLOP                        258
AlasanResign                           66
TanggalLahir                            0
TanggalHiring                           0
TanggalPenilaianKaryawan          

- Pada kolom 'SkorKepuasanPegawai' terdapat 5 data yang missing value. 
- Pada kolom 'JumlahKeikutsertaanProjek' terdapat 3 data yang missing value. 
- Pada kolom 'JumlahKeterlambatanSebulanTerakhir' terdapat 1 data yang missing value. 
- Pada kolom 'JumlahKetidakhadiran' terdapat 6 data yang missing value. 
- Pada kolom 'IkutProgramLOP' terdapat 258 data yang missing value.
- Pada kolom 'AlasanResign' terdapat 66 data yang missing value.

In [117]:
# Pengecekan data yang duplicate
print("\nPengecekan Data Duplicate:")
df.duplicated().value_counts()


Pengecekan Data Duplicate:


False    287
Name: count, dtype: int64

- Tidak ada data yang duplicate.

In [118]:
# Pengecekan jumlah unique value untuk kolom categorical
for col in categorical_features:
    unique = len(df[col].unique())
    print(f"{col} : {unique}")

StatusPernikahan : 5
JenisKelamin : 2
StatusKepegawaian : 3
Pekerjaan : 14
JenjangKarir : 3
PerformancePegawai : 5
AsalDaerah : 5
HiringPlatform : 9
TingkatPendidikan : 3
PernahBekerja : 2
AlasanResign : 12


In [119]:
# Pengecekan kolom 'PernahBekerja'
result = df['PernahBekerja'].unique()
result

array(['1', 'yes'], dtype=object)

- Fitur PernahBekerja adalah fitur yang hanya memiliki satu unique value.

In [120]:

result = df.loc[df['IkutProgramLOP'].isna()]
result[['Username', 'EnterpriseID', 'Pekerjaan', 'SkorKepuasanPegawai', 'PernahBekerja', 'IkutProgramLOP', 'AlasanResign']]
# result

Username  EnterpriseID                      Pekerjaan  \
6            madZebra1        100309                   Data Analyst   
7     cynicalMandrill7        110500   Software Engineer (Back End)   
8         lyingMussel8        110533    Software Engineer (Android)   
9         enviousMoth4        110927    Software Engineer (Android)   
11      grizzledSnipe7        111354   Software Engineer (Back End)   
..                 ...           ...                            ...   
280       anxiousTeal2        111378   Software Engineer (Back End)   
281       anxiousWasp6        100628  Software Engineer (Front End)   
282        dopeySheep0        106034                  Data Engineer   
283  yearningPorpoise4        106254       Product Design (UI & UX)   
286       puzzledFish5        106214                   Data Analyst   

     SkorKepuasanPegawai PernahBekerja  IkutProgramLOP   AlasanResign  
6                    3.0             1             NaN  masih_bekerja  
7                    5.0             1             NaN  masih_bekerja  
8                    NaN             1             NaN  masih_bekerja  
9                    5.0             1             NaN  masih_bekerja  
11                   5.0             1             NaN  tidak_bahagia  
..                   ...           ...             ...            ...  
280                  5.0             1             NaN            NaN  
281                  5.0             1             NaN            NaN  
282                  5.0             1             NaN            NaN  
283                  5.0             1             NaN      jam_kerja  
286                  3.0             1             NaN            NaN  

[258 rows x 7 columns]

## 1.2 Handling Missing Value

In [121]:
df_clean = df.copy()

In [122]:
# Pengecekan data yang hilang
print("\nPengecekan Missing Values:")
df.isnull().sum()


Pengecekan Missing Values:


Username                                0
EnterpriseID                            0
StatusPernikahan                        0
JenisKelamin                            0
StatusKepegawaian                       0
Pekerjaan                               0
JenjangKarir                            0
PerformancePegawai                      0
AsalDaerah                              0
HiringPlatform                          0
SkorSurveyEngagement                    0
SkorKepuasanPegawai                     5
JumlahKeikutsertaanProjek               3
JumlahKeterlambatanSebulanTerakhir      1
JumlahKetidakhadiran                    6
NomorHP                                 0
Email                                   0
TingkatPendidikan                       0
PernahBekerja                           0
IkutProgramLOP                        258
AlasanResign                           66
TanggalLahir                            0
TanggalHiring                           0
TanggalPenilaianKaryawan          

In [123]:
result = df.loc[df['SkorKepuasanPegawai'].isna()]
result[['Username', 'EnterpriseID', 'SkorKepuasanPegawai','Pekerjaan']]

Username  EnterpriseID  SkorKepuasanPegawai  \
8         lyingMussel8        110533                  NaN   
25         needySyrup4        101226                  NaN   
55   importedApricots7        105820                  NaN   
122   puzzledBurritos7        111373                  NaN   
166       resolvedCod5        110380                  NaN   

                        Pekerjaan  
8     Software Engineer (Android)  
25                Product Manager  
55   Software Engineer (Back End)  
122   Software Engineer (Android)  
166      Product Design (UI & UX)

In [124]:
# Mengisi nilai null pada Kolom 'SkorKepuasanPegawai' dengan Mean
df_clean['SkorKepuasanPegawai'].fillna(df_clean['SkorKepuasanPegawai'].mode()[0], inplace=True)

# Mengisi nilai null pada Kolom 'JumlahKeikutsertaanProjek' dengan Mean
df_clean['JumlahKeikutsertaanProjek'].fillna(df_clean['JumlahKeikutsertaanProjek'].mode()[0], inplace=True)

# Mengisi nilai null pada Kolom 'JumlahKeterlambatanSebulanTerakhir' dengan Mean
df_clean['JumlahKeterlambatanSebulanTerakhir'].fillna(df_clean['JumlahKeterlambatanSebulanTerakhir'].mode()[0], inplace=True)

# Mengisi nilai null pada Kolom 'JumlahKetidakhadiran' dengan Mean
df_clean['JumlahKetidakhadiran'].fillna(df_clean['JumlahKetidakhadiran'].mode()[0], inplace=True)

# Mengisi nilai null pada Kolom 'AlasanResign' dengan 'tanpa_alasan'
df_clean['AlasanResign'].fillna('tanpa_alasan', inplace=True)

C:\Users\WINDOWS\AppData\Local\Temp\ipykernel_16068\3929609102.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean['SkorKepuasanPegawai'].fillna(df_clean['SkorKepuasanPegawai'].mode()[0], inplace=True)
C:\Users\WINDOWS\AppData\Local\Temp\ipykernel_16068\3929609102.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we ar

- Melakukan handling missing value pada fitur SkorKepuasanPegawai, JumlahKeikutsertaanProjek, JumlahKeterlambatanSebulanTerakhir, dan JumlahKetidakhadiran dengan mengisinya menggunakan modus.

- Melakukan handling missing value pada fitur tanpa_alasan dengan mengisi nilainya menjadi tanpa_alasan

In [125]:
result = df_clean.loc[df_clean['AlasanResign'].isna()]
result[['Username', 'EnterpriseID', 'SkorKepuasanPegawai','Pekerjaan']]

Empty DataFrame
Columns: [Username, EnterpriseID, SkorKepuasanPegawai, Pekerjaan]
Index: []

## 1.3 Drop Features

In [126]:
# Menghapus Fitur yang Tidak Diperlukan
drop_features = ['Username', 'EnterpriseID', 'PernahBekerja', 'IkutProgramLOP']

# df.drop(drop_features, axis=1, inplace=True)
df_clean = df_clean.drop(drop_features, axis=1, inplace=False)
print(f"Jumlah baris : {df.shape[0]}")
print(f"Jumlah kolom : {df.shape[1]}")

Jumlah baris : 287
Jumlah kolom : 25


- Fitur-fitur yang dihapus yakni fitur-fitur yang tidak diperlukan seperti 'Username', 'EnterpriseID', 'PernahBekerja', dan 'IkutProgramLOP'.